app/packs/src/apps/mydb/elements/details/samples/SampleDetails.js

```diff
@@ -739,59 +748,121 @@ export default class SampleDetails extends React.Component {
     )
   }
 
-  addManualCas(e) {
-    DetailActions.updateMoleculeCas(this.props.sample, e.value);
+  // addManualCas(e) {
+  //   DetailActions.updateMoleculeCas(this.props.sample, e.value);
+  // }
+
+  addHypenToCas(cas, boolean) {
+    const { sample } = this.state;
+    const filterCas = cas.replace(/-/g, '');
+    const firstHypenIndex = filterCas.length - 3;
+    const secondHypenIndex = filterCas.length - 1;
+    console.log(filterCas);
+    const firstPart = filterCas.substring(0, firstHypenIndex);
+    const secondPart = filterCas.substring(firstHypenIndex, secondHypenIndex);
+    const thirdPart = filterCas.substring(secondHypenIndex, filterCas.length);
+    const format = `${firstPart}-${secondPart}-${thirdPart}`;
+    if (boolean) {
+      sample.xref = { ...sample.xref, cas: format };
+      this.setState({ sample });
+    }
+    console.log(format);
+    const checkFormat = format.match(/([0-9]{2,7})-([0-9]{2})-[0-9]/);
+    console.log(checkFormat);
+    return checkFormat;
+  }
+
+  checkCasValidation(cas, boolean) {
+    this.setState({ validateCas: true });
+    const filterCas = cas.replace(/-/g, '');
+    if (cas && cas !== '') {
+      let final = 0;
+      const isnum = /^[0-9-]+$/.test(cas) && !(/^[-]+$/.test(cas));
+      let match;
+      if (filterCas.length >= 5 && isnum) {
+        const checkFormat = cas.match(/([0-9]{2,7})-([0-9]{2})-[0-9]/);
+        match = checkFormat || this.addHypenToCas(cas, boolean);
+        const digits = (match[1] + match[2]).split('').reverse();
+        let sum = 0;
+        for (let i = 0; i < digits.length; i += 1) {
+          sum += (i + 1) * parseInt(digits[i], 10);
+        }
+        final = sum % 10;
+      }
+      const valid = final == cas.slice(-1);
+      if (!valid && boolean) {
+        this.setState({ validateCas: false });
+      }
+      const casOrSmile = match ? match[0] : 'smile';
+      return casOrSmile;
+    }
+    return cas;
   }
 
   moleculeCas() {
-    const { sample, isCasLoading } = this.state;
+    const { sample, isCasLoading, validateCas } = this.state;
     const { molecule, xref } = sample;
     const cas = xref ? xref.cas : '';
     const casLabel = cas && cas.label ? cas.label : '';
     let casArr = [];
     if (molecule && molecule.cas) {
-      casArr = molecule.cas.map(c => Object.assign({ label: c }, { value: c }));
+      casArr = molecule.cas.map((c) => ({ label: c, value: c })).filter((c) => c.value !== null);
```

Comment:
- Could you please break lines after each step of the process
- Please rename c by a meaningful name

Reply:
done